types/transactions.go

```diff
@@ -143,17 +144,51 @@ type (
 // ID returns the id of a transaction, which is taken by marshalling all of the
 // fields except for the signatures and taking the hash of the result.
 func (t Transaction) ID() TransactionID {
-	return TransactionID(crypto.HashAll(
-		t.SiacoinInputs,
-		t.SiacoinOutputs,
-		t.FileContracts,
-		t.FileContractRevisions,
-		t.StorageProofs,
-		t.SiafundInputs,
-		t.SiafundOutputs,
-		t.MinerFees,
-		t.ArbitraryData,
-	))
+	var hash crypto.Hash
+	h := crypto.NewHash()
+	enc := encoding.NewEncoder(h)
+
+	for i := range t.SiacoinInputs {
+		t.SiacoinInputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.SiacoinOutputs)))
+	for i := range t.SiacoinOutputs {
+		t.SiacoinOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.FileContracts)))
+	for i := range t.FileContracts {
+		enc.Encode(t.FileContracts[i])
+	}
+	encoding.WriteInt(h, len((t.FileContractRevisions)))
+	for i := range t.FileContractRevisions {
+		enc.Encode(t.FileContractRevisions[i])
+	}
+	encoding.WriteInt(h, len((t.StorageProofs)))
+	for i := range t.StorageProofs {
+		enc.Encode(t.StorageProofs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundInputs)))
+	for i := range t.SiafundInputs {
+		enc.Encode(t.SiafundInputs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundOutputs)))
+	for i := range t.SiafundOutputs {
+		t.SiafundOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.MinerFees)))
+	for i := range t.MinerFees {
+		t.MinerFees[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.ArbitraryData)))
+	for i := range t.ArbitraryData {
+		encoding.WritePrefix(h, t.ArbitraryData[i])
+	}
+	encoding.WriteInt(h, len((t.TransactionSignatures)))
+	for i := range t.TransactionSignatures {
```

Comment:
I was skeptical of this; seems like an easy thing to optimize. I tried a microbenchmark:
```go
xs := make([]int, 1e6)
sum := 0
for i := range xs {
	sum += xs[i]
}

// vs.

xs := make([]int, 1e6)
sum := 0
for _, n := range xs {
	sum += n
}
```
The former does in fact run faster. The generated assembly is almost identical except for one section:
```
4f392c: LEAQ 0x1(DI), R8	      
4f3939: MOVQ R8, DI	  

vs:

4f3930: MOVQ 0(AX), R9	        
4f3933: ADDQ $0x8, AX	         
4f3937: INCQ R8	         
```  
It appears that when using the index, the compiler uses `LEAQ` to load the value directly, whereas when using the value, it uses `ADDQ` and `INCQ` to calculate the memory offset. So the number of allocations and the number of copies are the same, but the latter is indeed marginally slower.